### ISSUE TITLE
Permit use of MSI to drive SYSCLK

### ISSUE THREAD

**Issue by thenewwazoo** [DoumanAsh/stm32l4x6_hal#1]
I'm not sure what this should look like (I'm still learning the concepts behind embedded-hal), but I'd like to have the ability to configure the MSI clock (MSIRANGE in RCC_CR), and use the LSE to calibrate MSI (MSIPLLEN in RCC_CR), as in 6.2.3 "Hardware auto calibration with LSE".

Right now, it looks like the default is to use either the PLL or HSI16 depending on the clock speed requested, right?

**Comment by DoumanAsh** [comment]
Well PLL/HSI are used as sources for system clock
By configuring MSI clock do you mean you'd like to use it as source. Afaik it should be possible.
CFGR's register SW allows you to set following clocks as sources:
```
Bits 1:0 SW[1:0]: System clock switch
Set and cleared by software to select system clock source (SYSCLK).
Configured by HW to force MSI oscillator selection when exiting Standby or Shutdown mode.
Configured by HW to force MSI or HSI16 oscillator selection when exiting Stop mode or in
case of failure of the HSE oscillator, depending on STOPWUCK value.
00: MSI selected as system clock
01: HSI16 selected as system clock
10: HSE selected as system clock
11: PLL selected as system cloc
```

But I'm not sure myself how it should look to be honest.
Configuration here is primarily inspired by one of the existing crates for other stm32 boards.

To be honest I did want to setup my board as quick as possible.
But I did not dive in details of clock configuration.
If you have knowledge on this topic I'd like to hear outline of possible configurations you can see as useful.
From here we could draw new API for configuration.

By itself clock configuration is not described in `embedded-hal` so afaik we can go as much as we'd like too.

**Comment by DoumanAsh** [comment]
Another thing to note: As clock is pretty vital we should ensure that it is impossible to misconfigure it

**Comment by thenewwazoo** [comment]
Right now, I'm thinking perhaps the `freeze` function could take an `Option<RtcClockType>`. If it's `Some(RtcClockType::LSE)`, then set `MSIPLLEN` and do some PLL math akin to the what's used to set sysclk.

**Comment by DoumanAsh** [comment]
Well `RtcClockType` is not really right enum to use as it specifically contains possible clocks for LCD's clock.

Wo could have separate method to calibrate MSI itself in `CFGR` and/or enable LSE clock(as it seems to require to be enabled as by default it is off? need to check it)

**Comment by DoumanAsh** [comment]
@thenewwazoo If you'd like to discuss things further it might be also convenient to talk over irc for example. You can find me on `#rust-embedded` channel or just PM me, my nick is `Douman`

**Comment by DoumanAsh** [comment]
MSI Clock initialization for reference from C HAL:
```c++
    /* Check the parameters */
    assert_param(IS_RCC_MSI(RCC_OscInitStruct->MSIState));
    assert_param(IS_RCC_MSICALIBRATION_VALUE(RCC_OscInitStruct->MSICalibrationValue));
    assert_param(IS_RCC_MSI_CLOCK_RANGE(RCC_OscInitStruct->MSIClockRange));

    /* When the MSI is used as system clock it will not be disabled */
    if((__HAL_RCC_GET_SYSCLK_SOURCE() == RCC_CFGR_SWS_MSI) )
    {
      if((READ_BIT(RCC->CR, RCC_CR_MSIRDY) != RESET) && (RCC_OscInitStruct->MSIState == RCC_MSI_OFF))
      {
        return HAL_ERROR;
      }

       /* Otherwise, just the calibration and MSI range change are allowed */
      else
      {
        /* To correctly read data from FLASH memory, the number of wait states (LATENCY)
           must be correctly programmed according to the frequency of the CPU clock
           (HCLK) and the supply voltage of the device. */
        if(RCC_OscInitStruct->MSIClockRange > __HAL_RCC_GET_MSI_RANGE())
        {
          /* First increase number of wait states update if necessary */
          if(RCC_SetFlashLatencyFromMSIRange(RCC_OscInitStruct->MSIClockRange) != HAL_OK)
          {
            return HAL_ERROR;
          }

          /* Selects the Multiple Speed oscillator (MSI) clock range .*/
          __HAL_RCC_MSI_RANGE_CONFIG(RCC_OscInitStruct->MSIClockRange);
          /* Adjusts the Multiple Speed oscillator (MSI) calibration value.*/
          __HAL_RCC_MSI_CALIBRATIONVALUE_ADJUST(RCC_OscInitStruct->MSICalibrationValue);
        }
        else
        {
          /* Else, keep current flash latency while decreasing applies */
          /* Selects the Multiple Speed oscillator (MSI) clock range .*/
          __HAL_RCC_MSI_RANGE_CONFIG(RCC_OscInitStruct->MSIClockRange);
          /* Adjusts the Multiple Speed oscillator (MSI) calibration value.*/
          __HAL_RCC_MSI_CALIBRATIONVALUE_ADJUST(RCC_OscInitStruct->MSICalibrationValue);

          /* Decrease number of wait states update if necessary */
          if(RCC_SetFlashLatencyFromMSIRange(RCC_OscInitStruct->MSIClockRange) != HAL_OK)
          {
            return HAL_ERROR;
          }          
        }

        /* Update the SystemCoreClock global variable */
        SystemCoreClock = HAL_RCC_GetSysClockFreq() >> AHBPrescTable[(RCC->CFGR & RCC_CFGR_HPRE)>> POSITION_VAL(RCC_CFGR_HPRE)];
        
        /* Configure the source of time base considering new system clocks settings*/
        HAL_InitTick (TICK_INT_PRIORITY);
      }
    }
    else
    {
      /* Check the MSI State */
      if(RCC_OscInitStruct->MSIState != RCC_MSI_OFF)
      {
        /* Enable the Internal High Speed oscillator (MSI). */
        __HAL_RCC_MSI_ENABLE();

        /* Get timeout */
        tickstart = HAL_GetTick();

        /* Wait till MSI is ready */
        while(READ_BIT(RCC->CR, RCC_CR_MSIRDY) == RESET)
        {
          if((HAL_GetTick() - tickstart) > MSI_TIMEOUT_VALUE)
          {
            return HAL_TIMEOUT;
          }
        }
         /* Selects the Multiple Speed oscillator (MSI) clock range .*/
        __HAL_RCC_MSI_RANGE_CONFIG(RCC_OscInitStruct->MSIClockRange);
         /* Adjusts the Multiple Speed oscillator (MSI) calibration value.*/
        __HAL_RCC_MSI_CALIBRATIONVALUE_ADJUST(RCC_OscInitStruct->MSICalibrationValue);

      }
      else
      {
        /* Disable the Internal High Speed oscillator (MSI). */
        __HAL_RCC_MSI_DISABLE();

        /* Get timeout */
        tickstart = HAL_GetTick();

        /* Wait till MSI is ready */
        while(READ_BIT(RCC->CR, RCC_CR_MSIRDY) != RESET)
        {
          if((HAL_GetTick() - tickstart) > MSI_TIMEOUT_VALUE)
          {
            return HAL_TIMEOUT;
          }
        }
      }
    }
  }
```

**Comment by thenewwazoo** [comment]
If you want to see my current thinking for the clock tree API, check out the [composable-clocks](https://github.com/thenewwazoo/stm32l4x6_hal/tree/composable-clocks) branch I'm working on. I need to refactor it a bit to use your `power.rs` work and do a little more testing but it appears to be working so far.

**Comment by thenewwazoo** [comment]
I don't know if you'd want to merge it in before I'm able to write clock reconfiguration (which might be a while, since I'm not doing it right now).

**Comment by DoumanAsh** [comment]
We can have clock reconfiguration separately I suppose so feel free to make PR

**Comment by DoumanAsh** [comment]
Since github forgot, Fixed by #4
